server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -295,11 +301,29 @@ private void processIndexShardSnapshots(ClusterChangedEvent event) {
                                 }
                             }
                         }
+                    } else {
+                        final ImmutableOpenMap<ShardId, ShardSnapshotStatus> clusterStateShards = entry.shards();
+                        final Snapshot snapshot = entry.snapshot();
+                        final Set<ShardId> cachedFailedShards = remoteFailedShardsCache.getOrDefault(snapshot, Collections.emptySet());
```

Comment:
I'm confused a bit by how the cache here works (in particular the clean-up). Maybe extract the caching infrastructure in ShardStateAction (in a separate PR), and just reuse that?

Reply:
I tried to make it work analogous to the way we handle org.elasticsearch.snapshots.SnapshotShardsService#shardSnapshots (rebuild the correct map from what the state seen in this method and replace the instance), but yea it's confusing => drying this up against the logic in `ShardStateAction` sounds good imo :)